django/db/models/fields/related.py

```diff
@@ -879,6 +882,57 @@ def _check_on_delete(self):
                     id='fields.E321',
                 )
             ]
+        elif on_delete in [DB_CASCADE, DB_RESTRICT, DB_SET_NULL] and (
+            any(  # generic relation
+                hasattr(field, 'bulk_related_objects') for field in
+                self.model._meta.private_fields
+            ) or
+            any(  # generic foreign key
+                hasattr(field, 'get_content_type') for field in
+                self.model._meta.private_fields
+            )
+        ):
+            return [
+                checks.Error(
+                    'Field specifies unsupported on_delete=DB_* on model '
+                    'declaring a GenericForeignKey.',
+                    hint='Change the on_delete rule.',
+                    obj=self,
+                    id='fields.E345',
+                )
+            ]
+        elif on_delete in [DB_CASCADE, DB_RESTRICT, DB_SET_NULL] and len(
+            # multi table inheritance
+            self.model._meta.concrete_model._meta.parents.values()
+        ):
+            return [
+                checks.Error(
+                    'Field specifies unsupported on_delete=DB_* on multi-table'
+                    ' inherited model.',
+                    hint='Change the on_delete rule.',
+                    obj=self,
+                    id='fields.E345',
+                )
+            ]
+        elif on_delete in [CASCADE, RESTRICT, SET_DEFAULT, SET_NULL] and (
+            any(  # field_A <- CASCADE <- field_B <- DB_CASCADE <- field_C
+                getattr(field_B.remote_field, 'on_delete', None) and
+                getattr(field_B.remote_field, 'on_delete') in [
+                    DB_CASCADE, DB_RESTRICT, DB_SET_NULL
+                ]
+                for field_B in self.remote_field.model._meta.fields
+            )
```

Comment:
First step is to just loosen this condition to not choke when `_meta` doesn't exist.